modules/core/src/test/java/org/apache/ignite/cache/ReleaseSegmentOnHistoricalRebalance.java

```diff
@@ -0,0 +1,408 @@
+     *
+     * @throws Exception If failed.
+     */
+    @Test
+    public void testReleaseDuringRebalanceIterator() throws Exception {
+        checkHistoricalRebalance(n -> {
+            IgniteInternalCache cachex = n.cachex(DEFAULT_CACHE_NAME);
+
+            GridCacheOffheapManager spy = spy(offheapMgr(cachex));
+
+            doAnswer(m -> {
+                WALPointer reserved = dbMgr(n).latestWalPointerReservedForPreloading();
+
+                assertNotNull(reserved);
+
+                Object o = m.callRealMethod();
+
+                release(n, reserved);
+
+                assertTrue(segmentAware(n).minReserveIndex(Long.MAX_VALUE));
+
+                return o;
+            }).when(spy).rebalanceIterator(any(), any());
+
+            offheapMgr(cachex, spy);
+        });
+    }
+
+    /**
+     * Checks that if the reservation is released immediately,
+     * there will be no errors and the rebalance will be completed.
+     *
+     * @throws Exception If failed.
+     */
+    @Test
+    public void testImmediateReleaseSegment() throws Exception {
+        checkHistoricalRebalance(n -> {
+            SegmentAware spy = spy(segmentAware(n));
+
+            doAnswer(m -> {
+                Object o = m.callRealMethod();
+
+                spy.release(m.getArgument(0));
+
+                return o;
+            }).when(spy).reserve(anyLong());
+
+            segmentAware(n, spy);
+        });
+    }
+
+    /**
+     * Checks that that if there is no reservation,
+     * there will be no errors and the rebalance will be completed.
+     *
+     * @throws Exception If failed.
+     */
+    @Test
+    public void testNoReserveSegment() throws Exception {
+        checkHistoricalRebalance(n -> {
+            SegmentAware spy = spy(segmentAware(n));
+
+            when(spy.reserve(anyLong())).thenAnswer(m -> false);
+
+            segmentAware(n, spy);
+        });
+    }
+
+    /**
+     * Populate the cache.
+     *
+     * @param cache Cache.
+     * @param cnt Entry count.
+     * @param o Key offset.
+     * @throws Exception If failed.
+     */
+    private void populate(IgniteCache<? super Object, ? super Object> cache, int cnt, int o) throws Exception {
+        for (int i = 0; i < cnt; i++) {
+            if (i % 100 == 0)
+                forceCheckpoint();
+
+            cache.put(i + o, new byte[64 * 1024]);
+        }
+    }
+
+    /**
+     * Set the spy to {@code CheckpointMarkersStorage#cpHistory}.
+     *
+     * @param n Node.
+     * @param spy Spy.
+     */
+    private void cpHistory(IgniteEx n, CheckpointHistory spy) {
+        CheckpointMarkersStorage s = getFieldValue(dbMgr(n), "checkpointManager", "checkpointMarkersStorage");
+
+        setFieldValue(s, "cpHistory", spy);
+    }
+
+    /**
+     * Set the spy to {@code GridCacheSharedContext#dbMgr}.
+     *
+     * @param n Node.
+     * @param spy Spy.
+     */
+    private void dbMgr(IgniteEx n, GridCacheDatabaseSharedManager spy) {
```

Comment:
Please do not abbreviate method names https://cwiki.apache.org/confluence/display/IGNITE/Coding+Guidelines#CodingGuidelines-Abbreviations